EIPS/eip-7709.md

```diff
@@ -0,0 +1,138 @@
+| `FORK_TIMESTAMP`          | TBD                                          |
+| `HISTORY_STORAGE_ADDRESS` | `0x25a219378dad9b3503c8268c9ca836a52427a4fb` |
+| `BLOCKHASH_SERVE_WINDOW`  | `256`                                        |
+| `HISTORY_SERVE_WINDOW`    | `8192`                                       |
+
+The `BLOCKHASH` opcode semantics remains the same as before. From the `fork_block` (first block where `block.timestamp >= FORK_TIMESTAMP`), the `BLOCKHASH` instruction should be updated to retrieve the requested blockhash from the contract storage.
+
+### Contract Implementation
+
+Exact evm assembly that can be used for the contract to resolve `BLOCKHASH`
+
+```
+// check if input > 8 byte value and revert if this isn't the case
+// the check is performed by comparing the biggest 8 byte number with
+// the call data, which is a right-padded 32 byte number.
+push8 0xffffffffffffffff
+push0
+calldataload
+gt
+push1 0x39
+jumpi
+
+// check if input > blocknumber-1 then return 0
+push1 0x1
+number
+sub
+push0
+calldataload
+gt
+push1 0x31
+jumpi
+
+// check if blocknumber > input + 8192 then return 0, no overflow expected for input of 8 bytes
+push0
+calldataload
+push2 0x2000
+add
+number
+gt
+push1 0x31
+jumpi
+
+// mod 8192 and sload
+push2 0x1FFF
+push0
+calldataload
+and
+sload
+
+// load into mem and return 32 bytes
+push0
+mstore
+push1 0x20
+push0
+return
+
+// return 0
+jumpdest
+push0
+push0
+mstore
+push1 0x20
+push0
+return
+
+// revert
+jumpdest
+push0
+push0
+revert
+
+stop
+```
+
+Note that the input contract read `32` bytes input as `calldataload`. Users and clients doing EVM call to this contract should left pad the `arg` correctly.
+
+<!-- TODO: bytecode is based off on first version and will be updated once assembly is locked down as it changes contract sender and address -->
+
+Corresponding bytecode:
+`60203611603157600143035f35116029575f356120000143116029576120005f3506545f5260205ff35b5f5f5260205ff35b5f5ffd00`
+
+#### Deployment
+
+For the fork_block (blocks where `block.timestamp >= FORK_TIMESTAMP`), we consider that the fork (`FORK_TIMESTAMP`) is already in effect for at least `HISTORY_SERVE_WINDOW` blocks (or from genesis if `fork_block < HISTORY_SERVE_WINDOW`). The following changes take effect:
+
+- `BLOCKHASH` SHOULD retrieve blockhashes from the system contract storage and MUST provide the storage access in the Verkle execution witness
+- The gas cost of the `BLOCKHASH` opcode is updated as per specification
+
+### [EIP-158](./eip-158.md) handling
+
+### Gas costs
+
+At FORK_TIMESTAMP, the corresponding storage access witness for the blockhash `SLOAD` lookups must also included in the execution witness, along with their access gas charges.
+
+## Rationale
+
+## Backwards Compatibility
+
+This EIP introduces a significant increase in the cost of `BLOCKHASH`, which could break use-cases that rely on the previous gas cost.
+
+## Test Cases
+
+TBD
```

Comment:
There should be at least the following tests cases:

* if `BLOCKHASH` is not called in the block, and neither is the contract what does the witness look like?
* if `BLOCKHASH` is called but not the contract, what does the witness look like?
* if the contract is called but not `BLOCKHASH`, what does the witness look like?

witness and gas costs, that is.

Reply:
also, if for whatever reason a testnet has prague and osaka within 8192 blocks, what are the returned values.

Reply:
The old spec was 8192, with backfill, until we changed it after seeing how invasive the changes would have been.

Reply:
I don't think the case where prague and osaka happen within 8192 blocks is as concerning as the case where prague and osaka happen within 256 blocks. With the former, the storage access just returns 0, with a proof of absence. This is not breaking because this is consistent with what happens at prague time (only the block hashes from `prague_fork_block - 1` are stored, all previous ones would return 0). However, if the osaka happens within 256 blocks, then we have break in functionality because:
- With EIP-2935 (at prague), we still serve the last 256 blocks from client state as we did in the past (not breaking)
- With EIP-7709 (at osaka), we now expect the block hashes to be retrieved through storage, meaning that blocks before `prague_fork_block - 1` would return `0`. This is breaking unless the Osaka fork happens at least 256 blocks before. Obviously won't be the case on Mainnet, but could certainly cause issues for testnets.

Reply:
Update with the suggested examples from @gballet , and added a mention of the breaking change if there is not enough blocks between prague and osaka.

Reply:
I only expect this in the case where it is prague at genesis.  We will not be testing transitions from devnets starting at Cancun.  In the case of prague at genesis all available slots for BLOCKHASH will be loaded from genesis.,even if it's less than 256.  But it will be all blocks on the chain.

Reply:
> There should be at least the following tests cases:
> 
> 
> 
> * if `BLOCKHASH` is not called in the block, and neither is the contract what does the witness look like?
> 
> * if `BLOCKHASH` is called but not the contract, what does the witness look like?
> 
> * if the contract is called but not `BLOCKHASH`, what does the witness look like?
> 
> 
> 
> witness and gas costs, that is.

Reply:
> for whatever reason a testnet has prague and osaka within 8192 blocks,


this EIP assumes that 2935 has been sufficiently activated ahead > 256 (so not even 8192), so adding a test case if that happens